### PR TITLE
Add command replay during outage

### PR DESCRIPTION
**Pull request by xpatiate**
Changes to read data from a stored optimisation plan and publish commands:
- add config value for the command payload action
- move MQTT publishing (for commands and errors) into separate package
- during outage, clean up stored plan to remove expired intervals on each pass
- during outage, select first current interval from plan and publish
- add unit conversion to ensure we always publish in kW
- add tests

### CODE REVIEW

**Comment by xpatiate** [comment]
The decline in coverage seems to be mainly due to error conditions, so I can add some more tests which trigger those

File: internal/plan/plan.go [hunk 1]
```diff
@@ -5,10 +5,14 @@ import (
...
 	"log/slog"
 	"os"
+	"sync"
+	"time"
 )
 
 type PlanHandler struct {
-	Path string `required:"True"`
+	logger *slog.Logger
+	mu     *sync.Mutex
```
Comment (serroba) [inline internal/plan/plan.go:14]: can you explain to me, why is this needed?
also, why is this a pointer to a mutex rather than the plain instance?

File: internal/plan/plan.go [hunk 2]
```diff
@@ -40,12 +44,25 @@ type OptimisationValue struct {
 	Unit  int     `json:"unit"`
 }
 
-func (o OptimisationPlan) IsEmpty(logger *slog.Logger) bool {
+func (o OptimisationPlan) IsEmpty() bool {
```
Comment (serroba) [inline internal/plan/plan.go:47]: I like you removed the logger. Is a better signature.
This change, however, make me realise, I don't think we are running the linting rules on this project, are we? if the logger wasn't used before, the linter rule could have complained about it.

File: internal/plan/plan.go [hunk 3]
```diff
@@ -63,10 +80,14 @@ func (p PlanHandler) ReadPlan() (OptimisationPlan, error) {
 }
 
 func (p PlanHandler) WritePlan(optPlan OptimisationPlan) error {
+	p.mu.Lock()
```
Comment (serroba) [inline internal/plan/plan.go:83]: a mutex helps us when we deal with concurrency and race conditions.
Why does the planner needs to worry about that? couldn't the planner be thought as a single threaded, non-concurrent type of application?

Comment (serroba) [reply internal/plan/plan.go:83]: Mind due, I am not implying what you have done here is wrong at all. I am just asking to learn why we need this.

File: internal/standby/standby.go [hunk 4]
```diff
@@ -107,15 +112,27 @@ func (s *Service) runDetector(ctx context.Context) {
...
 	timeSinceLastCmd := currentTime.Sub(s.storageSvc.GetCommandTimestamp())
-	s.logger.Debug("checking", "time since last command", timeSinceLastCmd)
+	s.logger.Debug("checking", "time since last command", timeSinceLastCmd, "current mode", s.getMode(), "currentTime", currentTime)
 	if timeSinceLastCmd > outageThreshold {
 		if s.InStandbyMode() {
 			s.logger.Info("Outage detected", "config threshold", outageThreshold, "time since last command", timeSinceLastCmd)
 			s.setMode(CommandMode)
 		} else {
 			s.logger.Debug("Ongoing outage", "config threshold", outageThreshold, "time since last command", timeSinceLastCmd)
+			err := s.planHandler.TrimPlan(currentTime)
```
Comment (serroba) [inline internal/standby/standby.go:122]: I am not sure, why do we need to trim the plan?

File: internal/standby/standby.go [hunk 5]
```diff
@@ -107,15 +112,27 @@ func (s *Service) runDetector(ctx context.Context) {
...
+			}
+		}
+		currentInterval, err := s.planHandler.GetCurrentInterval(currentTime)
+		if err != nil {
+			s.publisher.PublishError("getting current command", err)
+			return
+		}
+		err = s.publisher.PublishCommand(currentInterval)
+		if err != nil {
+			s.publisher.PublishError("publishing current command", err)
```
Comment (serroba) [inline internal/standby/standby.go:134]: Do you think we will be publishing back to MQTT if there is an issue?
I have a feeling we probably will want to get back to the cloud details about the outage when it happens and what things we have done during the outage.

Just open questions here, not a call for further change in current PR

File: internal/plan/plan.go (hunk 2 above)
Comment (xpatiate) [reply internal/plan/plan.go:47]: Ugh, I've been faithfully running `golangci-lint run` locally before every commit, and we also have it in CI, but I don't have a settings file for it 🤦‍♀️

**Review by serroba — APPROVED** [review]
Only a few open questions. No specific request for change though

File: internal/standby/standby.go (hunk 4 above)
Comment (xpatiate) [reply internal/standby/standby.go:122]: I felt like this would be a good idea to start with, then wasn't so sure it was useful by the time I had done it. My thinking was basically that if we are repeatedly reading from a stored plan, removing expired intervals would make it quicker to get to the current one. If you're doing that on every iteration though you are probably spending more time doing the cleanup than you would be just skipping past the stale intervals.

File: internal/plan/plan.go (hunk 1 above)
Comment (xpatiate) [reply internal/plan/plan.go:14]: The pointer is there to avoid an error from `govet` about copying lock by value: 
https://groups.google.com/g/golang-nuts/c/o-2N5wc8JcI

File: internal/plan/plan.go (hunk 3 above)
Comment (xpatiate) [reply internal/plan/plan.go:83]: I thought it would be safer to lock and unlock the file given that it's been written and read by different parts of the system - the MQTT subscriber callback `handlePlanMessage` writes the plan to file, and `CheckForOutage` reads it from the ticker-based goroutine (and also rewrites it, if we keep the `TrimPlan` part)

File: internal/plan/plan.go (hunk 2 above)
Comment (xpatiate) [reply internal/plan/plan.go:47]: I'll add that in a separate PR to avoid cluttering this one up with lint changes

File: internal/standby/standby.go (hunk 5 above)
Comment (xpatiate) [reply internal/standby/standby.go:134]: yeah I think the publisher will need to also record its activities to file, as presumably its MQTT messages aren't going to go very far during an outage. I'll add a ticket for that

File: internal/standby/standby.go (hunk 4 above)
Comment (serroba) [reply internal/standby/standby.go:122]: I feel that if it isn't needed, that this feature has come up out of necessity, I would probably drop it. It simplifies what we actively do and we may eventually have a different need that makes this unnecessary altogether

File: internal/plan/plan.go (hunk 1 above)
Comment (serroba) [reply internal/plan/plan.go:14]: Interesting

File: internal/plan/plan.go (hunk 2 above)
Comment (serroba) [reply internal/plan/plan.go:47]: sure thing

File: internal/plan/plan.go (hunk 3 above)
Comment (serroba) [reply internal/plan/plan.go:83]: 👍 probably makes sense then

File: internal/standby/standby.go (hunk 4 above)
Comment (jlordiales) [reply internal/standby/standby.go:122]: yeah, in practice our optimisation plans are not so big that just loading the whole thing into memory and iterating through it becomes a concern

Comment (xpatiate) [reply internal/standby/standby.go]: great, will remove

File: internal/plan/plan.go [hunk 6]
```diff
@@ -40,15 +44,28 @@ type OptimisationValue struct {
...
+	isAfterStart := targetTime.Equal(intStart) || targetTime.After(intStart)
+	isBeforeEnd := targetTime.Before(intEnd)
+	return isAfterStart && isBeforeEnd
+}
+
+func NewHandler(logger *slog.Logger, path string) Handler {
+	return Handler{logger: logger, mu: new(sync.RWMutex), Path: path}
+}
+
+func (p Handler) ReadPlan() (OptimisationPlan, error) {
```
Comment (jlordiales) [inline internal/plan/plan.go:68]: should this use the `RLock` method of the mutex?

Comment (xpatiate) [reply internal/plan/plan.go:68]: yes it should, thanks for pointing that out!

File: internal/plan/plan.go [hunk 7]
```diff
@@ -5,10 +5,14 @@ import (
...
+	"sync"
+	"time"
 )
 
-type PlanHandler struct {
-	Path string `required:"True"`
+type Handler struct {
+	logger *slog.Logger
+	mu     *sync.RWMutex
+	Path   string `required:"True"`
```
Comment (esimonov) [inline internal/plan/plan.go:15]: 1. What is `required:"True"` for? I can't see how this struct is being read to (and in case I'm missing where this happens, I would suggest to use a different struct for that - mixing configuration objects with working entities may be confusing).

2. Can we please keep the fields private? I thought we agreed on that approach when discussing the previous PR.

File: internal/publisher/publisher.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	Category  string    `json:"category"`
+	Message   string    `json:"message"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func (s *Service) PublishError(message string, receivedError error) {
+	s.logger.Error(message, "error", receivedError)
+
+	payload := ErrorPayload{
+		Category:  "Standby",
```
Comment (esimonov) [inline internal/publisher/publisher.go:43]: Creating constants for possible values here may be beneficial.

File: internal/publisher/publisher.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		Message:   fmt.Sprintf("Error %s: %s", message, receivedError),
+		Timestamp: time.Now(),
+	}
+	encPayload, err := json.Marshal(payload)
+	if err != nil {
+		s.logger.Error("marshalling error payload", "error", err)
```
Comment (esimonov) [inline internal/publisher/publisher.go:49]: Don't we want to return here? We'll end up Publishing nil if we don't.

File: internal/publisher/publisher.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	if s.cfg.MQTT.ErrorTopic == "" {
+		s.logger.Error("no error topic configured")
```
Comment (esimonov) [inline internal/publisher/publisher.go:53]: Return?

File: internal/publisher/publisher.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	return nil
+}
+
+func BuildCommandPayload(action string, optInterval plan.OptimisationInterval) CommandPayload {
+	meterValue := float64(optInterval.MeterPower.Value)
+	meterUnit := optInterval.MeterPower.Unit
+
+	var publishMeterValue float64
+	switch {
+	case meterUnit == 1:
```
Comment (esimonov) [inline internal/publisher/publisher.go:81]: Why do we have to use such encoding? Won't named constants be better?

File: internal/plan/plan.go (hunk 7 above)
Comment (xpatiate) [reply internal/plan/plan.go:15]: 1. you're right, this is not needed
2. My instinct was to make this one public - for some reason I feel that if a value is provided by the caller to initialise an object, it's only fair that the caller should be able to get that value back again (probably overloading the terms 'public' and 'private' with too much socioeconomic baggage) - but yeah, it doesn't need to be, will change.

File: internal/publisher/publisher.go (hunk 11 above)
Comment (xpatiate) [reply internal/publisher/publisher.go:81]: This is based on the schema used to translate the optimisation plan into a command message, which normally happens within the cloud orchestration service. We're reproducing that logic here (the standby service receives the optimisation plan and needs to turn it into a command message) so just have to work with the encoding that exists.

Comment (esimonov) [reply internal/publisher/publisher.go:85]: I meant declaring something like

```go
const (
        MeterPowerUnitMegaWatt = 1
        MeterPowerUnitKiloWatt = 2
        MeterPowerUnitWatt = 3
)
```

That could read a bit friendlier in the switch.

File: internal/publisher/publisher.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+	Value  float64 `json:"value"`
+}
+
+type ErrorPayload struct {
+	Category  string    `json:"category"`
+	Message   string    `json:"message"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+const errorCategory = "Standby"
```
Comment (esimonov) [inline internal/publisher/publisher.go:39]: I believe it should be `errorCategoryStandby` - to make it more obvious how to extend these values in the future.

File: internal/publisher/publisher.go (hunk 11 above)
Comment (xpatiate) [reply internal/publisher/publisher.go:85]: Ah got it, ok will do that

File: internal/publisher/publisher.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	return nil
+}
+
+func BuildCommandPayload(action string, optInterval plan.OptimisationInterval) CommandPayload {
+	meterValue := float64(optInterval.MeterPower.Value)
+	meterUnit := optInterval.MeterPower.Unit
+
+	var publishMeterValue float64
+	switch {
+	case meterUnit == MeterPowerUnitWatt:
```
Comment (esimonov) [inline internal/publisher/publisher.go:91]: Shouldn't Watt and MegaWatt be used the other way around? I may be misinterpreting this, but looks like we're dividing already small values, and multiplying already large values.

**Comment by esimonov** [comment]
I'm approving, but please take care of those missing returns.

**Review by esimonov — APPROVED** [review]

File: internal/publisher/publisher.go (hunk 13 above)
Comment (esimonov) [reply internal/publisher/publisher.go:91]: Nevermind - your calcs are correct.
